Merge 'refactoring/epoll_handler' into 'master'.

acatangiu opened the pull request:
## Reason for This PR

Brings all the decoupling work related to `Firecracker v1.0` to the master branch.

Fixes #1427 
Fixes #1428 
Fixes #1429 
Fixes #1430 
Fixes #1525 
Fixes #1538 
Fixes #1563 

## Description of Changes

### Merge method used

Rebased the `refactoring/epoll_handler` branch on top of `master`, this PR will do a clean merge back into `master`.

### High-level summary

#1417 Implement an I/O EventManager ("polly" crate).
#1489 MicroVM lifetime clearly split between pre-boot and post-boot.
#1490 Vmm interface organized in stacked layers each with its own concern.
#1494 Metrics driver and Serial device driven by EventManager (polly).
#1499 Block device driven by EventManager (polly).
#1518 Net device driven by EventManager (polly).
#1550 Vmm-Exit event and Api events driven by EventManager (polly).
#1575 Vsock device driven by EventManager (polly).

### High-level microVm interface

![vmm-components](https://user-images.githubusercontent.com/31917973/71740384-b5355500-2e64-11ea-9770-4f3496dab525.png)

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided.
- [x] The description of changes is clear and encompassing.
- [x] Either no docs need to be updated as part of this PR, or, the required
      doc changes are included in this PR. Docs in scope are all `*.md` files
      located either in the repository root, or in the `docs/` directory.
- [x] Code-level documentation for touched code is included in this PR.
- [x] No API changes are included in this PR.
- [x] The changes in this PR have no user impact.
- [x] No new `unsafe` code has been added.

acatangiu commented on src/vmm/src/lib.rs:
```diff
@@ -1646,1702 +415,45 @@ impl Vmm {
-
-    #[test]
-    // Test that loading the initrd is successful on different archs.
-    fn test_load_initrd() {
```
Good catch!
Since these tests were left out as a rebase error/miss, I will add them to the `add builder tests` commit in this PR.

aghecenco commented on src/arch/src/lib.rs:
```diff
@@ -52,6 +52,7 @@ pub enum DeviceType {
 }
 
 /// Type for passing information about the initrd in the guest memory.
+#[derive(Clone)]
```
Unnecessary, please remove

aghecenco commented on src/devices/src/legacy/mod.rs:
```diff
@@ -14,4 +14,5 @@ pub use self::i8042::Error as I8042DeviceError;
 pub use self::i8042::I8042Device;
 #[cfg(target_arch = "aarch64")]
 pub use self::rtc_pl031::RTC;
+pub use self::serial::ReadableFd;
```
nit
```suggestion
pub use self::serial::{ReadableFd, Serial};
```

aghecenco commented on src/devices/src/virtio/block/device.rs:
```diff
@@ -0,0 +1,948 @@
+        self.device_activated = device_activated;
+    }
+
+    fn activate(&mut self, _mem: GuestMemoryMmap) -> ActivateResult {
```
Shouldn't the `fn activate`s have been removed from all devices?

acatangiu replied:
The `vsock` device still uses it. I've opened #1620 that tracks unifying the activation mechanism.

aghecenco commented on src/devices/src/virtio/block/event_handler.rs:
```diff
@@ -0,0 +1,51 @@
+        let source = event.fd();
+        let event_set = event.event_set();
+
+        // TODO: also check for errors. Pending high level discussions on how we want
```
I saw this comment before. Is there a tracking issue?

acatangiu replied:
#1627

aghecenco commented on src/devices/src/virtio/block/mod.rs:
```diff
@@ -0,0 +1,37 @@
+pub const QUEUE_SIZES: &[u16] = &[QUEUE_SIZE];
+
+#[derive(Debug)]
+pub enum Error {
```
nit: please sort members alphabetically

aghecenco commented on src/devices/src/virtio/block/mod.rs:
```diff
@@ -0,0 +1,37 @@
+pub const NUM_QUEUES: usize = 1;
+pub const QUEUE_SIZES: &[u16] = &[QUEUE_SIZE];
+
+#[derive(Debug)]
```
Would be nice to `impl Display` for these (and other) errors originating in devices, that make their way to the end user. Out of scope for now, but I think we should have it.

acatangiu replied:
Let's do that after the merge.

aghecenco commented on src/devices/src/virtio/mmio.rs:
```diff
@@ -1,15 +1,14 @@
-// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
I don't think we should change the date on source files that were there before.

aghecenco commented on src/devices/src/virtio/mmio.rs:
```diff
@@ -429,551 +332,552 @@ impl BusDevice for MmioDevice {
-        assert_eq!(dummy_dev.avail_features(), 4);
-    }
-}
+// #[cfg(test)]
```
I'd rather remove the unit tests altogether, this is a lot of commented code.

acatangiu replied:
There's a pending PR #1609 that will make them work.

I propose we keep it like so when re-enabling tests, the diff will also show the changes made to the tests against the originals, should make it easier to review and/or check no tests are missed.

aghecenco commented on src/devices/src/virtio/vsock/device.rs:
```diff
@@ -23,62 +23,182 @@
+    /// to the backend for processing. Return `true` if descriptors have been added to the used
+    /// ring, and `false` otherwise.
+    pub fn process_tx(&mut self) -> bool {
+        debug!("vsock: epoll_handler::process_tx()");
```
The `epoll_handler` bit should be removed from the debug message

aghecenco commented on src/devices/src/virtio/vsock/device.rs:
```diff
@@ -222,7 +309,7 @@ mod tests {
         ];
         assert_eq!(ctx.device.device_type(), uapi::VIRTIO_ID_VSOCK);
-        assert_eq!(ctx.device.queue_max_sizes(), defs::QUEUE_SIZES);
+        // assert_eq!(ctx.device.queue_max_sizes(), defs::QUEUE_SIZES);
```
For small commented snippets, it would be helpful to leave a comment explaining why they don't run, and where the problem is tracked

acatangiu commented:
> It looks like not all the changes from f546253 ended up in this rebase. Specifically, we are loosing the ability to specify a different seccomp filter for VMM and for the vcpus.

What didn't make it exactly? We should be able to specify different filters for each thread now (we still specify the same cloned filter-set, but the capability for independent sets should be there).

acatangiu commented on src/devices/src/virtio/vsock/event_handler.rs:
```diff
@@ -0,0 +1,550 @@
+/// In a nutshell, the `EpollHandler` logic looks like this:
+/// - on TX queue event:
+///   - fetch all packets from the TX queue and send them to the backend; then
+///   - if the backend has queued up any incoming packets, fetch them into any available RX buffers.
```
I think the original is correct, as I understand it, the backend is checked for incoming packets on all events, and it always pushes them (the packets) into RX buffers.

iulianbarbu replied:
I guess I have to look at it less specific. Indeed, RX buffers are sufficient when working with network related packets/frames. Behind the scenes, these RX buffers (which are actually packets to be formed) are based on  `DescriptorChain`s  (HEAD and BUF/DATA descriptors) from RX virtio queue.

andreeaflorescu commented:
In Firecracker master we use different exit codes for bad configuration and invalid json. In this merge the error codes are unified and we just use bad configuration.

I assume it is easier to always exit with bad configuration since `VmResources::from_json` can now exit with error when the json is invalid or when configuring the resources failed. In case this change is intentional, should we also update the changelog to specify that an invalid configuration will cause Firecracker to exit with 151 as opposed to 152?

acatangiu commented:
@aghecenco @andreeaflorescu @iulianbarbu I've addressed all your comments, please take a look.

andreeaflorescu commented on src/devices/Cargo.toml:
```diff
@@ -5,12 +5,15 @@ authors = ["The Chromium OS Authors"]
 virtio_gen = { path = "../virtio_gen" }
+
+[dev-dependencies]
+vmm-sys-util = ">=0.3.1"
```
This looks like a leftover. We should use the local `utils` crate instead which is already declared as a dependency. Otherwise we might end up with multiple versions of vmm-sys-util.

The same comment applies to the vmm crate as well.

acatangiu replied:
fixed

andreeaflorescu commented:
> > It looks like not all the changes from [f546253](https://github.com/firecracker-microvm/firecracker/commit/f5462533a4e6919ecba4480ea62ff76fd8df8d94) ended up in this rebase. Specifically, we are loosing the ability to specify a different seccomp filter for VMM and for the vcpus.
> 
> What didn't make it exactly? We should be able to specify different filters for each thread now (we still specify the same cloned filter-set, but the capability for independent sets should be there).

We synced on this offline. On the master branch the `Vmm` would allow setting different seccomp filters on the vmm and vcpu threads through its public interface. The functionality was called through the public function `start_microvm` which received as parameters 2 filters (one for vmm and one for the vcpus). Since `start_microvm` no longer exists and the `start_vcpu` function that allows setting up custom seccomp filters is private, setting up different seccomp filters outside of the `Vmm` is not possible. Since there are other interface changes required, we'll do these all at once in a future PR.

acatangiu commented:
> We synced on this offline. On the master branch the Vmm would allow setting different seccomp filters on the vmm and vcpu threads through its public interface. The functionality was called through the public function start_microvm which received as parameters 2 filters (one for vmm and one for the vcpus). Since start_microvm no longer exists and the start_vcpu function that allows setting up custom seccomp filters is private, setting up different seccomp filters outside of the Vmm is not possible. Since there are other interface changes required, we'll do these all at once in a future PR.

I've added a commit to make the Vmm programmatic interface `public`.

aghecenco commented on src/vmm/src/lib.rs:
```diff
@@ -1231,405 +356,23 @@ impl Vmm {
+    #[cfg(target_arch = "x86_64")]
+    #[allow(unused)]
+    /// Returns the number of dirty pages in this microVM's memory.
+    pub fn get_dirty_page_count(&mut self) -> usize {
```
This has been removed from master in the meantime (https://github.com/firecracker-microvm/firecracker/commit/dd27c233d052a290e2a2ac54047bc276033c0e5a), thanks @andreeaflorescu for pointing it out.

acatangiu replied:
The function itself is correct and functioning and will be required going forward (even if right now is unused). I don't see why we should remove it just to add it back a bit later.

The dirty pages metric support has been indeed removed from both master and this PR.

aghecenco replied:
It was only used for the deleted metric. What will it be required for?

acatangiu replied:
You're right, the function gets the bitmap but only returns the `count`. We need a function that provides the bitmap.

I'll remove this to keep the merge as clean as possible. We'll add the correct function when we add its usecase.

andreeaflorescu commented on src/vmm/src/rpc_interface.rs:
```diff
@@ -0,0 +1,312 @@
+    /// Update the size of an existing block device specified by an ID. The ID is the first data
+    /// associated with this enum variant. This action can only be called after the microVM is
+    /// started.
+    RescanBlockDevice(String),
```
This action was removed in 53cf1bacadc23a21df361a894ac2887cafcb7139. We should also remove it from here.

acatangiu replied:
Good catch! This was removed from everywhere else, this is a leftover. Fixed.

acatangiu commented:
> Last part of the review. Sorry for the delay. It looks like we are missing one commit (the one that removes block device rescan).

We're not missing the commit, there was just a leftover action definition which I've now removed. Good catch! Fixed.
